cachi2/core/package_managers/yarn/project.py

```diff
@@ -301,15 +301,23 @@ class Project(NamedTuple):
     def is_zero_installs(self) -> bool:
         """If a project is using the zero-installs workflow or not.
 
-        This is determined by the existence of a non-empty yarn cache folder. For more details on
-        zero-installs, see: https://v3.yarnpkg.com/features/zero-installs.
+        This is determined either by the existence of a non-empty yarn cache folder
+        (default PnP mode) or by a presence of an expanded node_modules directory which would work
+        similarly or exactly the same way as with the NPM ecosystem.
+        For more details on zero-installs, see: https://v3.yarnpkg.com/features/zero-installs.
         """
-        dir = self.yarn_cache
+        node_linker = self.yarn_rc.node_linker
+        if node_linker is None or node_linker == "pnp":
+            if self.yarn_cache.path.exists() and self.yarn_cache.path.is_dir():
+                # in this case the cache folder will be populated with downloaded ZIP dependencies
+                return any(file.suffix == ".zip" for file in self.yarn_cache.path.iterdir())
```

Comment:
It might help improve the UX to have the error message list what we're checking for. If I'm just a casual or new user, it might not be clear to me from the error message why my request is being considered a zero-install since it could be unintentional

Reply:
> Something to consider for later: failing every request if the source code has something in .yarn/cache might affect a lot of users testing the tool locally, since Yarn populates it by default.

Yeah, good point. But that will be remediated by #388. Other than that, the cache doesn't need to be commited to the repo at all, the repo simply has to be clean of any artifacts, which doesn't cover unstaged changes, but does cover the cache. I think that's a reasonable premise for cachi2 to work.

Reply:
> Even running Cachi2 twice will cause the request to fail before https://github.com/containerbuildsystem/cachi2/pull/388 comes in

Why is that? (and here the credit needs to go to @chmeliik who pointed out that cachi2 will always fetch the dependencies to a different directory, so when we forbid zero installs that could not happen, could it?)

Reply:
So, we're setting two relevant yarnrc keys for the regular workflow: "[globalFolder](https://v3.yarnpkg.com/configuration/yarnrc#globalFolder)" and "[enableMirror](https://v3.yarnpkg.com/configuration/yarnrc#enableMirror)". Global folder is where the prefetched deps will be placed, and enableMirror just makes sure yarn will use the global folder as a proxy. 

The catch here is that we're respecting [cacheFolder](https://v3.yarnpkg.com/configuration/yarnrc#cacheFolder), that is what we're reading to figure out if a project is zero-installs or not. If we clone a non-zero-installs project, the cacheFolder will be empty, and the Cachi2 will succeed. But since globalFolder is only a proxy, the actual cacheFolder will be filled with zip files after the first processing, so Cachi2 will think it is a zero-installs project, and fail.

Reply:
> So, we're setting two relevant yarnrc keys for the regular workflow: "[globalFolder](https://v3.yarnpkg.com/configuration/yarnrc#globalFolder)" and "[enableMirror](https://v3.yarnpkg.com/configuration/yarnrc#enableMirror)".

Hmm, good point, which brings me to my next question: why are we even setting `enableMirror` to fetch dependencies? I mean, it makes sense for the actual container build, but we shouldn't need to mirror anything, sounds actually like a bug to me, IOW for the purpose of fetching deps we need to disable mirroring.

Reply:
> IOW for the purpose of fetching deps we need to disable mirroring

Set [enableGlobalCache](https://yarnpkg.com/configuration/yarnrc#enableGlobalCache): true instead? I haven't fully tested this yet, but I think that could work instead of populating both the local+global caches during the prefetch

Reply:
I tested that manually without cachi2, should be fine, I'll look into that in more detail, doesn't block this PR though.